pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,824 @@
+// has not been set.
+func (bc *BuildController) handleCompletedBuild(build *buildapi.Build, pod *kapi.Pod) (*buildUpdate, error) {
+	// Make sure that the completion timestamp has not already been set
+	if build.Status.CompletionTimestamp != nil {
+		return nil, nil
+	}
+
+	update := &buildUpdate{}
+	var podStartTime *metav1.Time
+	if pod != nil {
+		podStartTime = pod.Status.StartTime
+	}
+	setBuildCompletionTimestampAndDuration(build, podStartTime, update)
+
+	return update, nil
+}
+
+// updateBuild is the single place where any update to a build is done in the build controller.
+// It will check that the update is valid, peform any necessary processing such as calling HandleBuildCompletion,
+// and apply the buildUpdate object as a patch.
+func (bc *BuildController) updateBuild(build *buildapi.Build, update *buildUpdate, pod *kapi.Pod) error {
+
+	stateTransition := false
+	// Check whether we are transitioning to a different build phase
+	if update.phase != nil && (*update.phase) != build.Status.Phase {
+		stateTransition = true
+	} else if build.Status.Phase == buildapi.BuildPhaseFailed && update.completionTime != nil {
+		// Treat a failed->failed update as a state transition when the completionTime is getting
+		// updated. This will cause an event to be emitted and completion processing to trigger.
+		// We get into this state when the pod updates the phase through the build/details subresource.
+		// The phase, reason, and message are set, but no event has been emitted about the failure,
+		// and the policy has not been given a chance to start the next build if one is waiting to
+		// start.
+		update.setPhase(buildapi.BuildPhaseFailed)
+		stateTransition = true
+	}
+
+	if stateTransition {
+		// Make sure that the transition is valid
+		if !isValidTransition(build.Status.Phase, *update.phase) {
+			return fmt.Errorf("invalid phase transition %s -> %s", buildDesc(build), *update.phase)
+		}
+
+		// Log that we are updating build status
+		reasonText := ""
+		if update.reason != nil && *update.reason != "" {
+			reasonText = fmt.Sprintf(" ( %s )", *update.reason)
+		}
+
+		// Update build completion timestamp if transitioning to a terminal phase
+		if buildutil.IsTerminalPhase(*update.phase) {
+			var podStartTime *metav1.Time
+			if pod != nil {
+				podStartTime = pod.Status.StartTime
+			}
+			setBuildCompletionTimestampAndDuration(build, podStartTime, update)
+		}
+		glog.V(4).Infof("Updating build %s -> %s%s", buildDesc(build), *update.phase, reasonText)
+	}
+
+	// Ensure that a pod name annotation has been set on the build if a pod is available
+	if update.podNameAnnotation == nil && !common.HasBuildPodNameAnnotation(build) && pod != nil {
+		update.setPodNameAnnotation(pod.Name)
+	}
+
+	patchedBuild, err := bc.patchBuild(build, update)
+	if err != nil {
+		if stateTransition && *update.phase == buildapi.BuildPhasePending {
```

Comment:
ok so we explicitly don't want to retry this particular event due to this particular failure because it could lead to us creating 2 pods for the same build.  got it.  I think it definitely merits a code comment, pending whether we rework things or not.